ion-schema/src/resolver/resolve_fn.rs

```diff
@@ -0,0 +1,563 @@
+                    //    to borrow the schema and type names.
+                    //  * The solution is interior mutability, but we don't want type references to use
+                    //    Cell because then they are not Sync, and we don't want to use Mutex because
+                    //    TypeReferences are in the hot path for validation.
+                    //  * Thankfully, we don't need the TypeReferences to have interior mutability all
+                    //    the time. It's only necessary for resolving (and un-resolving) the type references.
+                    //  * So instead, we just have a limited scope for treating this as immutable.
+                    //
+                    // We could also work around this by creating a "WhileBeingResolvedSchema" that
+                    // uses `Cell` to hold `TypeDefinition`s, but that's a lot of boilerplate code.
+                    unsafe { as_mutable(tr).resolve_type_coordinates(Some(tc)) }
+                }
+                Err(e) => type_resolution_errors.push(e),
+            }
+        });
+    }
+
+    if !type_resolution_errors.is_empty() {
+        // TODO: Report multiple errors
+        IonSchemaResult::Err(type_resolution_errors.into_iter().next().unwrap())?;
+    }
+
+    let schemas = schemas.into_iter().collect();
+
+    let schema_store = Arc::new(SchemaStore { schemas });
+
+    Ok(schema_idx_by_name
+        .into_iter()
+        .map(|(name, index)| (name, ResolvedSchema::new(schema_store.clone(), index)))
+        .collect())
+}
+
+/// Un-resolves the schemas and returns ownership of the underlying [`SchemaDocument`]s to the caller,
+/// in a map of `schema_id` to `SchemaDocument`.
+///
+/// If there are any un-dropped `ResolvedSchema` or `ResolvedType` that depend on the underlying
+/// `SchemaDocument`s that are _not_ passed into this function, the function will return `None`.
+pub fn unresolve(
+    schemas: HashMap<String, ResolvedSchema>,
+) -> Option<HashMap<String, SchemaDocument>> {
+    if schemas.is_empty() {
+        return Some(HashMap::new());
+    }
+
+    let (_, any_resolved_schema) = schemas.iter().next().unwrap();
+    let schema_store = any_resolved_schema.schema_store();
+
+    drop(schemas);
+
+    let unresolved_schemas = Arc::into_inner(schema_store)?
+        .schemas
+        .into_iter()
+        .map(|(name, schema)| {
+            schema.walk(&mut |tr: &TypeReference| {
+                // SAFETY: See note in [`resolve`]. Justification is the same.
+                unsafe { as_mutable(tr).resolve_type_coordinates(None) }
+            });
+            (name, schema)
+        })
+        .collect();
+    Some(unresolved_schemas)
+}
+
+/// Cast an immutable reference to a mutable reference.
+#[allow(clippy::mut_from_ref)]
+unsafe fn as_mutable<T: ?Sized>(val: &T) -> &mut T {
+    (val as *const T as *mut T).as_mut().unwrap_unchecked()
+}
```

Comment:
This is undefined behavior. To do this without causing UB, you need to wrap the reference that you'll be manually toggling between `&` and `&mut` inside an [`UnsafeCell`](https://doc.rust-lang.org/std/cell/struct.UnsafeCell.html).

Reply:
If I put it in an `UnsafeCell`, then types and schemas are no longer `Sync`, which would be undoing #151.

I think I can come up with an alternative solution.

Reply:
There's a lot of `clone()`-ing of schema and type names, but I have update this so that now it doesn't use unsafe code, and it doesn't require the use of any kind of `Cell`.